### PR TITLE
add package service links and clean up RUNID usage

### PR DESCRIPTION
**Pull request by pdowler**
add package service links when datalink finds multiple non-preview downloads
use Traceable tagging interface to determine when to append RUNID to generated download URLs

### CODE REVIEW

File: caom2-datalink-server/src/main/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessor.java [hunk 1]
```diff
@@ -226,6 +242,23 @@ else if (ReleaseType.META.equals(a.getReleaseType()))
                 }
             }
         }
+        log.warn("numFiles: " + numFiles);
```
Comment (brianmajor) [inline caom2-datalink-server/src/main/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessor.java]: is the WARN level intentional?

File: caom2-datalink-server/src/main/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessor.java [hunk 2]
```diff
@@ -179,6 +192,9 @@ else if (ReleaseType.META.equals(a.getReleaseType()))
                 readable = ar.metaReadable;
             // else: new releaseType is not likely without major caom design change
             
+            if (readable != null) {
+                pkgReadable = pkgReadable && readable;
```
Comment (brianmajor) [inline caom2-datalink-server/src/main/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessor.java:196]: Don't think the pkgReadable in the condition is necessary but does no harm.  (ie could be: pkgReadable = readable)

File: caom2-datalink-server/src/main/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessor.java [hunk 3]
```diff
@@ -226,6 +242,23 @@ else if (ReleaseType.META.equals(a.getReleaseType()))
...
+                    link.url = getPackageURL(pkg, uri);
+                    link.contentType = CONTENT_TYPE_TAR;
+                    link.description = "single download containing all files (previews and thumbnails excluded)";
+                    link.readable = pkgReadable;
+                } catch (MalformedURLException ex) {
+                    link.errorMessage = "failed to create package link: " + ex;
+                }
+                ret.add(link);
+            }
+        }
```
Comment (brianmajor) [inline caom2-datalink-server/src/main/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessor.java:261]: Do you need to add the link to ret when numFiles == 1?

File: caom2-datalink-server/src/test/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessorTest.java [hunk 4]
```diff
@@ -93,15 +93,17 @@
 
     static
     {
-        Log4jInit.setLevel("ca.nrc.cadc", Level.INFO);
+        Log4jInit.setLevel("ca.nrc.cadc.caom2", Level.DEBUG);
```
Comment (brianmajor) [inline caom2-datalink-server/src/test/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessorTest.java]: should be set back to INFO

**Review by brianmajor — CHANGES_REQUESTED** [review]

File: caom2-datalink-server/src/main/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessor.java (hunk 1 above)
Comment (pdowler) [reply caom2-datalink-server/src/main/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessor.java]: no. fixing :-)

File: caom2-datalink-server/src/main/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessor.java (hunk 2 above)
Comment (pdowler) [reply caom2-datalink-server/src/main/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessor.java:196]: The idea here is that once pkgReadable become false (non readable artifact) it stays false.

File: caom2-datalink-server/src/main/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessor.java (hunk 3 above)
Comment (pdowler) [reply caom2-datalink-server/src/main/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessor.java:261]: no, that link var is the pkg link in the block and we only want a pkg link with multiple files; the individual links are always added up at line ~206

File: caom2-datalink-server/src/test/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessorTest.java (hunk 4 above)
Comment (pdowler) [reply caom2-datalink-server/src/test/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessorTest.java]: fixed

File: caom2-datalink-server/src/main/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessor.java (hunk 2 above)
Comment (brianmajor) [reply caom2-datalink-server/src/main/java/ca/nrc/cadc/caom2/datalink/ArtifactProcessor.java:196]: Ah, yeah, missed the loop

**Review by brianmajor — APPROVED** [review]

**Pull request merged**